File: example-crs-webservice/crs-userspace/libs/libMSA/cpp/tests/cpu/test_affinity.cpp
#include "libmsa/cpu/affinity.h"
#include <gtest/gtest.h>
#include <pthread.h>
#include <sched.h>
#include <unistd.h>
#include <vector>

std::vector<int> get_process_affinity() {
    cpu_set_t cpuset;
    CPU_ZERO(&cpuset);

    if (sched_getaffinity(0, sizeof(cpu_set_t), &cpuset) == -1) {
        throw std::runtime_error("Failed to get CPU affinity");
    }

    std::vector<int> cpus;
    for (int i = 0; i < CPU_SETSIZE; ++i) {
        if (CPU_ISSET(i, &cpuset)) {
            cpus.push_back(i);
        }
    }
    return cpus;
}

TEST(CpuAffinityTest, SetCpuAffinity) {
    int num_cpus = sysconf(_SC_NPROCESSORS_ONLN);
    ASSERT_GT(num_cpus, 1);

    int num_tests = 10;
    for (int i = 0; i < num_tests; ++i) {
        int target_cpu = rand() % num_cpus;
        std::vector<int> target_cpus = {target_cpu};

        CpuAffinity::setCpuAffinity(target_cpus);

        std::vector<int> process_affinity = get_process_affinity();
        ASSERT_EQ(process_affinity, target_cpus);
    }
}

int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: example-crs-webservice/crs-userspace/libs/libMSA/python/tests/cpu/test_affinity.py
import unittest
import psutil
import random
import os

from libmsa import set_cpu_affinity

class TestSetCpuAffinity(unittest.TestCase):
    def test_set_cpu_affinity(self):
        self.num_tests = 10

        available_cpus = list(range(psutil.cpu_count()))
        print(f'Number of available cpus: {len(available_cpus)}')
        print(f'Cpu: {available_cpus}')

        for _ in range(self.num_tests):
            if len(available_cpus) < 2:
                self.skipTest("Not enough CPUs available to test CPU affinity.")
            
            target_cpus = [random.randint(0, len(available_cpus)  - 1)]
            set_cpu_affinity(target_cpus)
            
            process = psutil.Process(os.getpid())
            self.assertEqual(process.cpu_affinity(), target_cpus, "CPU affinity for process not set correctly.")

            for thread in process.threads():
                thread_process = psutil.Process(thread.id)
                self.assertEqual(thread_process.cpu_affinity(), target_cpus, f"CPU affinity for thread {thread.id} not set correctly.")
        

if __name__ == '__main__':
    unittest.main()